### PR TITLE
Revert "types: allow passing generic attributes to elements"

### PR DESCRIPTION
**Pull request by kumilingus**
Reverts clientIO/joint#1731

### CODE REVIEW

**Comment by kumilingus** [comment]
Hey @danielpza, I will have to revert 2 of your PRs unfortunately. They both introduce a breaking change (it wasn't caught because Travis CI was disabled).

The problem is that due to the class parameters inference, it's not possible to do the following.
```ts
const rect = new shapes.standard.Rectangle({ myAttribute: 1 });
```

It will technically result in:
```ts
class Rectangle extends dia.Element<{ myAttribute: 1 }> {}
```

 Which makes a call like this invalid:
```ts
rect.set({ position: { x: 10, y: 10 }}); 
```

I was trying to find any solution to this, but didn't succeed ([StackOverflow Question](https://stackoverflow.com/questions/73556247/generic-constraints-are-ignored-when-inferring-class-parameters)).

File: types/joint.d.ts [hunk 1]
```diff
@@ -1996,7 +1996,7 @@ export namespace shapes {
 
         type RectangleAttributes = dia.Element.GenericAttributes<RectangleSelectors>;
 
-        class Rectangle<A = RectangleAttributes> extends dia.Element<A> {
+        class Rectangle extends dia.Element<RectangleAttributes> {
```
Comment (danielpza) [inline types/joint.d.ts:1999]: Would something like this work?
```suggestion
         class Rectangle<A extends RectangleAttributes = RectangleAttributes> extends dia.Element<A> {
```

Comment (danielpza) [reply types/joint.d.ts:1999]: Or perhaps

```suggestion
        class Rectangle<A = RectangleAttributes> extends dia.Element<A & RectangleAttributes> {
```

**Comment by danielpza** [comment]
Perhaps this could work https://www.typescriptlang.org/play?#code/C4TwDgpgBAsiAq5oF4oG8oEMD8AuKAdgK4C2ARhAE4A0UZehpFlUAvgNwBQnAxgDaYAzoNggAwgOEAeeFAgAPYBAIATEXESQoqDUgB86TlGNQeAewKDglIj2BnKACmD54ASnSsjJkhGAALMxVnfAB5EgBLYBlaAGsIEDMAM1FNCAMAMlSkDzQvL14LKygScUlBAEZtQggAd1EJIUEpXUg9RwxMfCrWNy5SxuEKgDpfAKCOum62Pu5Cy2ASsqaAJmqCOobyya6oHtmAegOoAHV-ECgIkQCruUpKB2xOI6hZQUCiPhU5RWVvgANWhB-nIAG7KKDjIgAc38kKQUFqQkuBCSVEoEG+SQeJCwpiK1ls9hYYEwlEwYyow04A3KK1GfkCwQwZGmvS4QA

File: types/joint.d.ts (hunk 1 above)
Comment (kumilingus) [reply types/joint.d.ts:1999]: The `npx grunt ts` test still fails. This does not stop the class parameter inference.

**Comment by kumilingus** [comment]
The `npx grunt ts` test would fail.

Plus I am not sure I want to change all the related methods because of this (`set`, `attr`, `prop`, and possible other). We don't know if it would introduce some other issues.

Especially when there is an easy workaround. Do not extend from `shapes.standard`, but define the link from scratch.

```ts
export class MyLink<T extends joint.dia.Link.Attributes> extends joint.dia.Link<T> {

    defaults() {
        // defaults() is not a function but an object hash in the `dia.Link`
        const linkDefaults = super.defaults as unknown as Partial<T>;
        return {
            ...linkDefaults,
            source: { x: 1, y: 1 },
            type: 'MyLink',
            attrs: {
                line: {
                    connection: true,
                    stroke: '#333333',
                    strokeWidth: 2,
                    strokeLinejoin: 'round',
                    targetMarker: {
                        'type': 'path',
                        'd': 'M 10 -5 0 0 10 5 z'
                    }
                },
                wrapper: {
                    connection: true,
                    strokeWidth: 10,
                    strokeLinejoin: 'round'
                }
            }
        };
    }

    preinitialize(): void {
        this.markup = [{
            tagName: 'path',
            selector: 'wrapper',
            attributes: {
                'fill': 'none',
                'cursor': 'pointer',
                'stroke': 'transparent',
                'stroke-linecap': 'round'
            }
        }, {
            tagName: 'path',
            selector: 'line',
            attributes: {
                'fill': 'none',
                'pointer-events': 'none'
            }
        }];
    }
}

interface MyAttributes extends joint.dia.Link.Attributes {
    attribute1: boolean;
}

class MyChildLink extends MyLink<MyAttributes>  {

    defaults() {
        return {
            ...super.defaults(),
            type: 'MyChildLink',
            attribute1: true
        };
    }
}
```

Note that I am ok to re-open those PRs and merge them in `v4.0`.

**Comment by danielpza** [comment]
I understand, I'll see if I can find a workaround

**Pull request merged**